src/proof/PGTransformationAlgorithms.cc

```diff
@@ -25,379 +25,324 @@ along with Periplo. If not, see <http://www.gnu.org/licenses/>.
+                        //Sufficient to propagate modifications via merge
+                        mergeClauses(n->getAnt1()->getClause(), n->getAnt2()->getClause(), n->getClause(), n->getPivot());
+                        for (clauseid_t clauseid : n->getResolvents()) {
+                            if (getNode(clauseid)) { q.push_back(clauseid); }
+                        }
+                        // NOTE extra check
+                        if (proofCheck() > 1) checkClause(n->getId());
+                    } else {
+                        assert(not piv_in_ant1 or not piv_in_ant2);
+                        //Second case: pivot not in ant1 or not in ant2
+                        //Remove resolution step, remove n, ant without pivots gains n resolvents
+                        bool choose_ant1 = false;
+                        //Pivot not in ant1 and not in ant2
+                        if (not piv_in_ant1 and not piv_in_ant2) {
+                            // Choose one of the two antecedents heuristically
+                            choose_ant1 = chooseReplacingAntecedent(n);
+                        } else {
+                            choose_ant1 = not piv_in_ant1 and piv_in_ant2;
+                        }
 
-				if(!isSetVisited1(id2))
-				{
-					mpz_init( safe_lit_set[id2] );
-					setVisited1(id2);
-					mpz_set(safe_lit_set[id2], incr_safe_lit_set_2);
-				}
-				else
-					mpz_and(safe_lit_set[id2], safe_lit_set[id2], incr_safe_lit_set_2);
-			}
-			// NOTE important, free memory for node id
-			mpz_clear( safe_lit_set[id] );
-		}
-		else if(n->isLeaf())
-		{
-			// NOTE important, free memory for node id
-			mpz_clear( safe_lit_set[id] );
-		}
-		else opensmt_error_();
-	}
+                        ProofNode * replacing = choose_ant1 ? n->getAnt1() : n->getAnt2();
+                        ProofNode * other = choose_ant1 ? n->getAnt2() : n->getAnt1();
 
-	delete(DFSvec_);
-	delete [] safe_lit_set;
-	mpz_clear(incr_safe_lit_set_1);
-	mpz_clear(incr_safe_lit_set_2);
-	resetVisited1();
+                        replacing->remRes(id);
+                        other->remRes(id);
 
-	if ( verbose() > 1 ) { std::cerr << "; Recycling end, restructuring begin" << std::endl; }
-	if ( verbose() > 1 )
-	{
-		uint64_t mem_used = memUsed();
-		reportf( "; Memory used after recycling, before restructuring: %.3f MB\n",  mem_used == 0 ? 0 : mem_used / 1048576.0 );
-	}
+                        auto const & resolvents = n->getResolvents();
+                        for (clauseid_t clauseid : resolvents) {
+                            assert(clauseid < getGraphSize());
+                            ProofNode * res = getNode(clauseid);
+                            assert(res);
+                            if (res->getAnt1() == n) { res->setAnt1(replacing); }
+                            else if (res->getAnt2() == n) { res->setAnt2(replacing); }
+                            else opensmt_error_();
```

Comment:
or `OsmtInternalException`?

Reply:
There is some shuffling  in the next PR that touches this code, so I think it's better to do this in the later PR.